File: week13/3/car_pooling.py
# You are given the integer capacity and an array `trips` where trips[i] = [numPassengers, from, to]
# indicates that the ith trip has `numPassengers` passengers, and the locations to pick them up
# and drop them off are `from` and `to`, respectively.
#
# The vehicle only drives east (i.e., it cannot turn around and drive west).
#
# Return True if it is possible to pick up and drop off all passengers for all the given trips, or False otherwise.
#
# Example 1:
# Input: trips = [[2, 1, 5], [3, 3, 7]], capacity = 4
# Output: False
#
# Example 2:
# Input: trips = [[2, 1, 5], [3, 3, 7]], capacity = 5
# Output: True
#
# Constraints:
# - 1 <= trips.length <= 1000
# - trips[i].length == 3
# - 1 <= numPassengers <= 100
# - 0 <= from < to <= 1000
# - 1 <= capacity <= 10^5


def bubble_sort(arr):
    """
    Custom bubble sort function to sort events. (ascending order)
    Sort by location first, and if locations are the same, sort by passenger change.
    """
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            # Compare current event and the next event
            # Sort by location; if equal, sort by passenger change
            if arr[j][0] > arr[j + 1][0] or (
                arr[j][0] == arr[j + 1][0] and arr[j][1] > arr[j + 1][1]
            ):
                # Swap arr[j] and arr[j + 1]
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

def carPooling(trips, capacity):
    """
    Determine if it is possible to pick up and drop off all passengers
    without exceeding the car's capacity.
    
    :param trips: List of trips where each trip is [numPassengers, from, to].
    :param capacity: Maximum number of passengers the car can hold.
    :return: True if all trips can be completed within capacity, False otherwise.
    """
    if len(trips) == 0:
        return True # no passengers exist
    
    # Extract Events:
    # For each trip [numPassengers, from, to], create two events:
    # Pickup at from (+numPassengers).
    # Drop-off at to (-numPassengers).
    events = []
    for trip in trips:
        numPassengers, pickUp, dropOff = trip
        events.append((pickUp, +numPassengers))
        events.append((dropOff, -numPassengers))

    # Sort Events:
    # Sort by location, and if two events are at the same location, prioritize drop-off over pickup.
    bubble_sort(events)

    # Simulate Passenger Flow:
    # Traverse the sorted events, maintain a running total of passengers, and check if it exceeds the capacity.
    curr = 0 # tracking the number of current passengers
    for event in events:
        location, passengerChange =  event
        curr += passengerChange
        # print("location: ", location, "curr: ", curr)
        if curr > capacity:
            return False
    
    # If we processed all events without exceeding capacity, return True
    return True


# Example test cases
def test_carPooling():
    assert carPooling([[2, 1, 5], [3, 3, 7]], 4) == False, "Test Case 1 Failed"
    assert carPooling([[2, 1, 5], [3, 3, 7]], 5) == True, "Test Case 2 Failed"
    assert carPooling([[2, 2, 5], [3, 5, 7]], 4) == True, "Test Case 3 Failed"
    assert carPooling([[3, 2, 8], [4, 4, 6], [10, 1, 3]], 11) == False, "Test Case 4 Failed"
    print("All Test Cases Passed!")

# Uncomment to test
# test_carPooling()

File: week13/3/car_pooling.c
// You are given the integer capacity and an array `trips` where trips[i] = {numPassengers, from, to}
// indicates that the ith trip has `numPassengers` passengers, and the locations to pick them up
// and drop them off are `from` and `to`, respectively.
//
// The vehicle only drives east (i.e., it cannot turn around and drive west).
//
// Return true if it is possible to pick up and drop off all passengers for all the given trips, or false otherwise.
//
// Example 1:
// Input: trips = {{2, 1, 5}, {3, 3, 7}}, capacity = 4
// Output: false
//
// Example 2:
// Input: trips = {{2, 1, 5}, {3, 3, 7}}, capacity = 5
// Output: true
//
// Constraints:
// - 1 <= trips.length <= 1000
// - trips[i].length == 3
// - 1 <= numPassengers <= 100
// - 0 <= from < to <= 1000
// - 1 <= capacity <= 10^5

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Custom swapping function for 2D arrays
void swap(int* a, int* b, int size) {
    for (int i = 0; i < size; i++) {
        int temp = a[i];
        a[i] = b[i];
        b[i] = temp;
    }
}

// Bubble sort function to sort events
void bubble_sort(int** arr, int size) {
    for (int i = 0; i < size - 1; i++) {
        for (int j = 0; j < size - i - 1; j++) {
            // Sort by location first; if equal, sort by passenger change
            if (arr[j][0] > arr[j + 1][0] || 
                (arr[j][0] == arr[j + 1][0] && arr[j][1] > arr[j + 1][1])) {
                swap(arr[j], arr[j + 1], 2); // Each event has two elements
            }
        }
    }
}

// Function to check if car pooling is possible
bool carPooling(int** trips, int tripsSize, int* tripsColSize, int capacity) {
    // Edge case: no trips
    if (tripsSize == 0) {
        return true;
    }

    // Step 1: Extract events
    int eventsSize = tripsSize * 2;
    int** events = (int**)malloc(eventsSize * sizeof(int*));
    for (int i = 0; i < eventsSize; i++) {
        events[i] = (int*)malloc(2 * sizeof(int)); // Each event has 2 elements
    }

    int index = 0;
    for (int i = 0; i < tripsSize; i++) {
        int numPassengers = trips[i][0];
        int pickUp = trips[i][1];
        int dropOff = trips[i][2];

        // Add pickup and drop-off events
        events[index][0] = pickUp;
        events[index][1] = numPassengers;
        index++;

        events[index][0] = dropOff;
        events[index][1] = -numPassengers;
        index++;
    }

    // Step 2: Sort events using bubble sort
    bubble_sort(events, eventsSize);

    // Step 3: Simulate passenger flow
    int currentPassengers = 0;
    for (int i = 0; i < eventsSize; i++) {
        currentPassengers += events[i][1];
        if (currentPassengers > capacity) {
            // Free memory before returning
            for (int j = 0; j < eventsSize; j++) {
                free(events[j]);
            }
            free(events);
            return false;
        }
    }

    // Free memory before returning
    for (int i = 0; i < eventsSize; i++) {
        free(events[i]);
    }
    free(events);
    return true;
}

// Example test cases
void test_carPooling() {
    // Test Case 1
    int trips1[][3] = {{2, 1, 5}, {3, 3, 7}};
    int* trips1Ptrs[] = {trips1[0], trips1[1]};
    int tripsColSize1[] = {3, 3};
    printf("%s\n", carPooling(trips1Ptrs, 2, tripsColSize1, 4) ? "True" : "False"); // Output: False

    // Test Case 2
    int trips2[][3] = {{2, 1, 5}, {3, 3, 7}};
    int* trips2Ptrs[] = {trips2[0], trips2[1]};
    int tripsColSize2[] = {3, 3};
    printf("%s\n", carPooling(trips2Ptrs, 2, tripsColSize2, 5) ? "True" : "False"); // Output: True

    // Test Case 3
    int trips3[][3] = {{2, 2, 5}, {3, 5, 7}};
    int* trips3Ptrs[] = {trips3[0], trips3[1]};
    int tripsColSize3[] = {3, 3};
    printf("%s\n", carPooling(trips3Ptrs, 2, tripsColSize3, 4) ? "True" : "False"); // Output: True

    // Test Case 4
    int trips4[][3] = {{3, 2, 8}, {4, 4, 6}, {10, 1, 3}};
    int* trips4Ptrs[] = {trips4[0], trips4[1], trips4[2]};
    int tripsColSize4[] = {3, 3, 3};
    printf("%s\n", carPooling(trips4Ptrs, 3, tripsColSize4, 11) ? "True" : "False"); // Output: False
}

int main() {
    test_carPooling();
    return 0;
}